edx-solutions/xblock-drag-and-drop-v2
# Prevent overlapping drop placements

**Description:** This PR updates item placement logic:

* Zone alignment "none" (aka no alignment) is removed, dropped items always use the "aligned" layout option rather than being placed where the learner dropped them.
* "Max items per zone" setting added to allow limiting max number of items dropped into a single zone - course authors can use this setting to provide some kind of a hint, or to prevent visual overflow when too many items are dropped into single zone.

**JIRA:** [SOL-1979](https://openedx.atlassian.net/browse/SOL-1979)

**Discussions:** [SOL-1979](https://openedx.atlassian.net/browse/SOL-1979) and https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/83#issuecomment-236017205

**Sandboxes:**

* LMS: http://dndv2-sandbox9.opencraft.hosting/
* Studio: http://studio-dndv2-sandbox9.opencraft.hosting/


**Note:** at the moment #86 was merged, but #87 was not, and this PR does not include changes from #87; as a result, DnDv2 on sandbox show "Submit" button, but it does not do anything - this is an expected behavior until #87 is merged

**Testing instructions:**

Items alignment:
1. Open new or existing DnDv2 block in studio, open editor, go to "Zones" tab. *Expected:* zone "Alignment" dropdown does not have "none" option; default is "center"
2. Open DnDv2 XBlock in LMS.
3. Make sure that items dropped into a zone are aligned according to zone alignment configuration.
4. Existing zones with "none" alignment should be treated as having "center"alignment; this should be the case even if an XBlock was not updated in Studio after this PR is applied.

Max items per zone:
1. Open new or existing DnDv2 block in studio, open editor, go to "Items" tab. *Expected:* "Max items per zone" is displayed.
2. Setting "max items per zone" to ~~zero or~~ empty string has no further effect (make sure you can put as much items into a single dropzone as you wish in LMS)
3. Setting "max items per zone" to zero or negative value ~~should cause validation errors~~ will result in value being set to empty string
4. ~~Setting "max items per zone" to positive value will validate item placement. If any zone has more "correct" items than "max items per zone" a validation error will be shown, and XBlock will not be updated.~~
5. After setting "max items per zone" to positive value ~~and passing the validation~~, publish the unit and open it in LMS. *Expected:* can only put up to "max items per zone" items into a single zone.
6. *Note:* existing answers are not updated, so it is possible to have more than "max items per zone" items dropped into a single zone if "max items" setting changed after the answer was submitted.

**Screenshots:**

Studio "Zones" tab:

![image](https://cloud.githubusercontent.com/assets/3330048/17551549/79ba91d6-5f03-11e6-93ce-72042a44b884.png)

Studio "Items" tab:

![image](https://cloud.githubusercontent.com/assets/3330048/17770812/67533cc8-6548-11e6-8aa6-c996b833ee53.png)





LMS - item alignment:

![image](https://cloud.githubusercontent.com/assets/3330048/17551802/d18e97d0-5f04-11e6-8ed2-467a9352ed05.png)

Attempting to exceed max items per zone:

![image](https://cloud.githubusercontent.com/assets/3330048/17551840/06e31b68-5f05-11e6-9f80-e39ae9363337.png)

**Reviewers:**
- [ ] ~~a11y: @cptvitamin~~  a11y review will be performed later, for all changes in v2.1: [SOL-1978](https://openedx.atlassian.net/browse/SOL-1978)

**Checklist:**
- [x] Internal review
- [x] Upstream review
